indico/modules/categories/controllers/display.py

```diff
@@ -612,11 +618,23 @@ def _find_nearest_category(self, category_chain):
         raise Exception(f'Category {self.category.id} not found in category chain')
 
     def _get_event_data(self, event_query):
+        def calculate_keyword_id(kw):
+            return crc32(kw) + 3
+
         data = []
         categories = {}
         rooms = {}
+        keywords = {}
+        allowed_keywords = event_settings.get('allowed_keywords')
+        allowed_keywords_set = set(allowed_keywords)
         tz = self.category.display_tzinfo
         for event in event_query:
+            event.keywords = [keyword for keyword in event.keywords if keyword in allowed_keywords_set]
+            for keyword in event.keywords:
+                keyword_id = calculate_keyword_id(keyword)
+                if keyword_id not in keywords:
+                    keywords[keyword_id] = {'id': keyword_id, 'title': keyword,
+                                            'color': f'#{generate_contrast_colors(keyword_id).background}'}
```

Comment:
I'd use `keywords.setdefault(keyword_id, {...})` here.

And since `generate_contrast_colors` is fully deterministic AFAICT (it ONLY depends on the seed value, right?) put `@functools.cache` on it so it's memoized.

Reply:
Both done.